### PR TITLE
Proposal for model_summary update

### PR DESCRIPTION
**Pull request by sebastienwood**
I did a bit of work on model.summary() as discussed in https://forums.fast.ai/t/developer-chat/22363/942?u=sebastienwood
Now the model name, optimizer, loss function and callbacks functions are listed as well. Extra code for AdamW. Would however require to further discuss what and whatnot include, and how to present it to the users.

### CODE REVIEW

File: fastai/callbacks/hooks.py [hunk 1]
```diff
@@ -181,6 +184,18 @@ def model_summary(m:Learner, n:int=70):
...
     res += f"Total non-trainable params: {total_params - total_trainable_params:,}\n"
+           
+    res += f"Optimized with {str(m.opt_func)[25:-1].replace('>', '')}\n"
+    if m.true_wd and "adam" in str(m.opt_func).lower():
+        res += f"Using true weight decay as discussed in https://www.fast.ai/2018/07/02/adam-weight-decay/ \n"
+    res += f"Loss function : {m.loss_func.__class__.__name__}\n"
+    res += "=" * n + "\n"
+    res += "Callbacks functions applied \n"
+    for cbk in m.callbacks:
+        if cbk.__doc__ is None:
```
Comment (sgugger) [inline fastai/callbacks/hooks.py]: I'd remove the calls to __doc__ and only print the names: there is a `doc` function that fastai provides that will show the doc if needed, with the links to the source.

File: fastai/callbacks/hooks.py [hunk 2]
```diff
@@ -181,6 +184,18 @@ def model_summary(m:Learner, n:int=70):
     res += f"\nTotal params: {total_params:,}\n"
     res += f"Total trainable params: {total_trainable_params:,}\n"
     res += f"Total non-trainable params: {total_params - total_trainable_params:,}\n"
+           
+    res += f"Optimized with {str(m.opt_func)[25:-1].replace('>', '')}\n"
+    if m.true_wd and "adam" in str(m.opt_func).lower():
```
Comment (sgugger) [inline fastai/callbacks/hooks.py]: Don't test for Adam here, the behavior is different for all optimizers, even SGD when there is momentum (as discussed in the same article).

**Comment by sgugger** [comment]
Thanks for tackling this! I left a few comments for changes I think would make it a little bit better.

File: fastai/callbacks/hooks.py (hunk 2 above)
Comment (sebastienwood) [reply fastai/callbacks/hooks.py]: Ok ! (I didn't understood that when reading the doc :\)

**Comment by sebastienwood** [comment]
I made a quick update following your comments, it may be worth to precise in the doc that true_wd has impacts outside of Adam ! :) Which makes me think : does the Learner wd parameter erase the opt_func wd parameter ? 
Eg in :
opt_func = partial(optim.SGD, lr=args.lr, momentum=args.mom, nesterov = args.nest, weight_decay=args.wd)
        learn = Learner(data, model, loss_func=nn.CrossEntropyLoss(), metrics=[accuracy], opt_func=opt_func, callback_fns=ShowGraph, true_wd=False)

I could precise it in summary() if it's the case :)

**Comment by sgugger** [comment]
Oh, if the docs are wrong, feel free to suggest a PR to fix them! If you pass a wd to learner, it will be passed at the creation of the optimizer, yes.

**Comment by sgugger** [comment]
Just restyled two lines. Are the TODO still relevant?

**Comment by sebastienwood** [comment]
You are right they're not anymore. I fixed a small bug and added the weight decay flag :)

**Comment by sgugger** [comment]
Looking great, thanks again!
